### PR TITLE
Managing archived data in block level

### PR DESCRIPTION
**Pull request by nekia**
#### Type of change

- New feature

#### Description

In current edition, archiving/discarding of block is controlled in blockfile based. This design was based on the assumption that the block data is identical across the peers. But it has figured out that this assumption is incorrect, we need to change the core logic into the block based control for block archiving to cover the wider conditions.

#### Additional details

We've been shared this new design at the fabric contributors meeting.
https://docs.google.com/presentation/d/1tzD6l8VwLUeDytYeEChqigsIaqMw4Z2NQj9pNah8Msc/edit?usp=sharing

### CODE REVIEW

File: common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go [hunk 1]
```diff
@@ -42,9 +43,9 @@ func newBlockfileArchiver(id string, mgr *blockfileMgr) *blockfileArchiver {
 	loggerArchive.Info("newBlockfileArchiver: ", id)
 
 	blockfileDir := filepath.Join(blockarchive.BlockStorePath, ChainsDir, id)
-	arch := &blockfileArchiver{id, mgr, blockfileDir, 1}
+	arch := &blockfileArchiver{id, mgr, blockfileDir, 1, 0}
 
-	if blockarchive.IsArchiver {
+	if blockarchive.IsArchiver || blockarchive.IsClient {
```
Comment (anand-jpa) [inline common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go:47]: Do we need "if" condition here?
Is there any other actor other than Archiver and Client?

File: common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go [hunk 2]
```diff
@@ -205,20 +178,12 @@ func (arch *blockfileArchiver) deleteArchivedBlockfile(fileNum int) error {
...
-	}
+func (arch *blockfileArchiver) isNeedArchiving() bool {
 
-	loggerArchive.Debug("There is no blockfile to be archived yet")
+	remoteArchivedBlockHeight := blockarchive.GossipService.ReadArchivedBlockHeight(common.ChannelID(arch.chainID))
+	localCurrentBlockHeigh := arch.mgr.getBlockchainInfo().Height
+	localDiscardedBlockHeight := arch.currDiscardedBlockHeigh
+	// numKeepLatestBlocks := blockarchive.NumKeepLatestBlocks
+	loggerArchive.Infof("local : %d  vs  archived : %d  vs  discarded : %d", localCurrentBlockHeigh, remoteArchivedBlockHeight, localDiscardedBlockHeight)
 	return false
```
Comment (anand-jpa) [inline common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go]: This function is always returning "false".
What is the purpose for this function?

File: common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go [hunk 3]
```diff
@@ -205,20 +178,12 @@ func (arch *blockfileArchiver) deleteArchivedBlockfile(fileNum int) error {
...
-	if len(files) > keepFileNum {
-		loggerArchive.Debugf("%d blockfile(s) should be archived", len(files)-keepFileNum)
-		return true
-	}
+func (arch *blockfileArchiver) isNeedArchiving() bool {
 
-	loggerArchive.Debug("There is no blockfile to be archived yet")
+	remoteArchivedBlockHeight := blockarchive.GossipService.ReadArchivedBlockHeight(common.ChannelID(arch.chainID))
+	localCurrentBlockHeigh := arch.mgr.getBlockchainInfo().Height
+	localDiscardedBlockHeight := arch.currDiscardedBlockHeigh
```
Comment (anand-jpa) [inline common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go]: Minor point - arch.currDiscardedBlockHeigh (t) is missing

File: common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go (hunk 1 above)
Comment (nekia) [reply common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go:47]: There is as-is role (peer that is neither Archiver nor Client).

File: common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go (hunk 2 above)
Comment (nekia) [reply common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go]: Sorry. This function has not been used any more. I delete it.

File: common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go (hunk 1 above)
Comment (anand-jpa) [reply common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go:47]: oh okay.

File: common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go (hunk 3 above)
Comment (nekia) [reply common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go]: This variable has not been used any more. I'll delete it

Comment (nekia) [reply common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go]: I've found similar typo at `currentArchivedBlockHeigh`. I'll fix it.

File: common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go [hunk 4]
```diff
@@ -35,83 +36,119 @@ const (
...
+// discardBlockfilelIfNecessary is called every time a blockfile is finalized (reached the maximum size of data chunk) on client node.
+// If there are enough amount of blockfiles on local file system to be discarded, the actual discarding routine will be triggered.
+func (arch *blockfileArchiver) discardBlockfilelIfNecessary() {
+
+	// Running in context of Client
+
+	loggerArchiveClient.Infof("discardBlockfilelIfNecessary [%s]", arch.chainID)
+
+}
+
```
Comment (anand-jpa) [inline common/ledger/blkstorage/fsblkstorage/blockfile_archiver.go:187]: This function is empty, just writing log entry.

**Review by anand-jpa — APPROVED** [review]
Looks good to me

**Pull request merged**